components/scream/cime_config/buildnml

```diff
@@ -72,20 +75,27 @@ def ordered_dump(data, item, Dumper=yaml.SafeDumper, **kwds):
 ###############################################################################
 def evaluate_selectors(element, case, ez_selectors):
 ###############################################################################
+    """
+    Evaluate and remove selectors from the unprocessed XML nml file in the repo.
+
+    Elements with selectors are removed. If the selector evaulates to True, then
+    the corresponding element text becomes the new text value of the original
```

Comment:
Does the script check that no two unprocessed XML entries have a selector evaluating to true? E.g, if `hgrid=ne4`, and `nlevs=72`

```
<my_value hgrid="ne4"> Jim </my_value>
<my_value nlevs="72"> Not Jim </my_value>
```
would buildnml spot the error or will it simply pick the last one that ends up evaluating to true? Or, worse, both?

Reply:
@bartgol . Good question. The current implementation is that multiple matches are allowed and the match that occurs latest has the highest precedence.

Reply:
To answer your question, "Not Jim" would be set for my_value.

Reply:
I think that's probably ok for now, so long as whoever adds new entries to the defaults.xml file pays good attention, ensuring that no duplicates are possible after selector resolution. But perhaps, since we have the chance to improving things, we want to be safe down the road, and either prevent this, or at least warn the user.

Reply:
Actually, warning the user might be not too hard, right? We can scan the current content of the fully resolved xml, and verify that the new entry is not there, no?

Tbc, I'm ok merging this without implementing this safeguard, but I do think we should add it at some point.

Reply:
We also need to keep in mind the behavior of eam's namelist XML system. I'm OK with diverging from them when it suits our needs, but I'd prefer to stay compatible with them if it's all the same to us. It's clear from looking at their file that it works the way I described above. For example, just looking within the top few lines of this file, I see:

```
<dtime dyn="fv"                    >1800</dtime>
<dtime dyn="fv"    waccmx="1">300 </dtime>
```

This pattern only makes sense if multiple matches are supported with the later match taking precedence. I suppose its also possible that precedence is derived from the number of selectors matched.

Reply:
I've always intuitively expected these entries to choose the most specific option (ie. the one with the most selectors). But I guess choosing the "most recent match" by just running through each line and reassigning the variable every time you have a match might be easier. If we establish an expectation that the simplest case will be first and the most complex last, most recent match could be ok.

Reply:
Ah, good point. There might be two selectors that both eval to true, but one is more specific than the other. We can live with that for now, but boy, that _does_ look like a source of sneaky bugs!

Unfortunately, our buildnml would become quite complicated if we had to keep track of "independent" duplicates. We'd have to store some extra lists, to check if one option is a "specialization" of the other, in which case, all is good, and possibly throw an error if two xml entries have selectors that eval to true, but are "independent".

Given that EAM still works the way it does, I _assume_ this scenario never happened, so I'm ok with living with it.

Reply:
Oof, and yet another idea: we could enforce that selections are more specific as additional selection elements are added:

```
<foo>default</foo>
<foo s1="a">1</foo>
<foo s1="a" s2="b">2</foo>
```

^^^ This would be ok since the selections are increasingly "specific" for each sequential selection element.

```
<foo>default</foo>
<foo s1="a" s2="b">2</foo>
<foo s1="a">1</foo>
```

^^^ This would be an error. Something like "specificity inversion for foo".